src/main/java/com/thealgorithms/datastructures/hashmap/hashing/HashMapCuckooHashing.java

```diff
@@ -0,0 +1,255 @@
+        if (hash < 0) {
+            hash += hsize;
+        }
+        return hash;
+    }
+
+    public int hashing2(int key) {
+        int hash = key / hsize;
+        hash %= hsize;
+        if (hash < 0) {
+            hash += hsize;
+        }
+        return hash;
+    }
+
+    /**
+     * inserts the key into the hash map by wrapping it as an Integer object, then uses while loop to insert new key
+     * if desired place is empty, return.
+     * if already occupied, continue while loop over the new key that has just been pushed out.
+     * if while loop continues more than Thresh, rehash table to new size, then push again.
+     *
+     * @param key the desired key to be inserted in the hash map
+     */
+
+    public void insertHash(int key) {
+        Integer wrappedInt = key, temp;
+        int hash, maxloop = 0;
+
+        if (isFull()) {
+            System.out.println("Hash table is full\n");
+            return;
+        }
+
+        if (contains(key)) {
+            System.out.println("Key already inside, no duplicates allowed\n");
+            return;
+        }
+
+        while (maxloop <= thresh) {
+            maxloop++;
+            hash = hashing1(key);
+
+            if (buckets[hash] == null || buckets[hash] == AVAILABLE) {
+                buckets[hash] = wrappedInt;
+                size++;
+                checkLoadFactor();
+                return;
+            }
+
+            temp = buckets[hash];
+            buckets[hash] = wrappedInt;
+            wrappedInt = temp;
+            hash = hashing2(temp);
+            if (buckets[hash] == null || buckets[hash] == AVAILABLE) {
+                buckets[hash] = wrappedInt;
+                size++;
+                checkLoadFactor();
+                return;
+            }
+
+            temp = buckets[hash];
+            buckets[hash] = wrappedInt;
+            wrappedInt = temp;
+        }
+        System.out.println("Need to rehash table\n");
+        reHashTable();
+        insertHash(key);
+    }
+
+    /**
+     * creates new HashMapCuckooHashing object, then inserts each of the elements in the previous table to it with its new hash functions.
+     * then refers current array to new table.
+     *
+     */
+    public void reHashTable() {
+        HashMapCuckooHashing newT = new HashMapCuckooHashing(hsize * 2);
+        for (int i = 0; i < hsize; i++) {
+            if (buckets[i] != null && buckets[i] != AVAILABLE) {
+                newT.insertHash(this.buckets[i]);
+            }
+        }
+        this.hsize *= 2;
+        this.buckets = newT.buckets;
+        this.thresh = (int) (Math.log(hsize) / Math.log(2)) + 2;
+    }
+
+
+    /**
+     * deletes a key from the hash map and adds an available placeholder
+     *
+     * @param key the desired key to be deleted
+     */
+    public void deleteHash(int key) {
+        Integer wrappedInt = key;
+        int hash = hashing1(key);
+        if (isEmpty()) {
+            System.out.println("Table is empty");
```

Comment:
Like here. You cannot delete `key` from the table if it's not there

Reply:
got it. fixed now.